src/Castle.Windsor/MicroKernel/Arguments.cs

```diff
@@ -17,211 +17,233 @@ namespace Castle.MicroKernel
+		/// <returns><see cref="Arguments"/></returns>
+		public static Arguments FromProperties(object instance)
 		{
-			private readonly List<IArgumentsComparer> nestedComparers = new List<IArgumentsComparer>();
+			var arguments = new Arguments();
+			arguments.InsertProperties(instance);
+			return arguments;
+		}
 
-			public ArgumentsComparerExtended(IEnumerable<IArgumentsComparer> customStores)
+		/// <summary>
+		/// A factory method that can be used to supply <see cref="Arguments"/> to <see cref="IKernel.Resolve{T}(Castle.MicroKernel.Arguments)"/>. Creates an <see cref="Arguments"/> instance from an <see cref="IReadOnlyDictionary{S, O}"/>.
+		/// </summary>
+		/// <param name="readOnlyDictionary"><see cref="IReadOnlyDictionary{S, O}"/> where the dictionary is shallow copied up front into <see cref="Arguments"/> and then used as dependencies</param>
+		/// <returns><see cref="Arguments"/></returns>
+		public static Arguments FromReadOnlyDictionary(IReadOnlyDictionary<string, object> readOnlyDictionary)
+		{
+			var arguments = new Arguments() { isReadOnly = false };
+			foreach (var entry in readOnlyDictionary)
 			{
-				nestedComparers = new List<IArgumentsComparer>(customStores);
+				arguments.Add(entry.Key, entry.Value);
 			}
 
-			public IArgumentsComparer[] CustomComparers
+			arguments.isReadOnly = true;
+			return arguments;
+		}
+
+		/// <summary>
+		/// A factory method that can be used to supply <see cref="Arguments"/> to <see cref="IKernel.Resolve{T}(Castle.MicroKernel.Arguments)"/>. Creates an <see cref="Arguments"/> instance from an array of objects by using their types as keys.
+		/// </summary>
+		/// <param name="typedArguments">The object type and instance are copied into <see cref="Arguments"/> where the values are used as dependencies</param>
+		/// <returns><see cref="Arguments"/></returns>
+		public static Arguments FromTypedArguments(params object[] typedArguments)
+		{
+			var arguments = new Arguments();
+			foreach (var @object in typedArguments)
 			{
-				get { return nestedComparers.ToArray(); }
+				if (@object == null)
+				{
+					throw new ArgumentNullException(nameof(typedArguments), 
+						"Given array has null values. Only non-null values can be used as arguments.");
+				}
+				arguments.Add(@object.GetType(), @object);
 			}
+			return arguments;
+		}
 
-			public override bool Equals(object x, object y)
+		private sealed class ArgumentsComparer : IEqualityComparer<object>
+		{
+			public new bool Equals(object x, object y)
 			{
-				foreach (var store in nestedComparers)
+				if (x is string a)
 				{
-					bool areEqual;
-					if (store.RunEqualityComparison(x, y, out areEqual))
-					{
-						return areEqual;
-					}
+					return StringComparer.OrdinalIgnoreCase.Equals(a, y as string);
 				}
-				return base.Equals(x, y);
+				return x == y;
```

Comment:
This change from `object.Equals` to `==` shouldn't break anything since our keys should always be types here, but if you weren't aware the `==` operator is overloaded not overridden in the class hierarchy, so since `x` and `y` are both defined as `object` it'll result in a reference equality which wouldn't work for a string (but we handle strings above).

See https://blogs.msdn.microsoft.com/csharpfaq/2004/03/29/when-should-i-use-and-when-should-i-use-equals/ if you want an explanation with example.